src/libcore/intrinsics.rs

```diff
@@ -962,59 +962,129 @@ extern "rust-intrinsic" {
     /// value is not necessarily valid to be used to actually access memory.
     pub fn arith_offset<T>(dst: *const T, offset: isize) -> *const T;
 
-    /// Copies `count * size_of<T>` bytes from `src` to `dst`. The source
-    /// and destination may *not* overlap.
+    /// Copies `count * size_of::<T>()` bytes from `src` to `dst`. The source
+    /// and destination must *not* overlap.
     ///
-    /// `copy_nonoverlapping` is semantically equivalent to C's `memcpy`.
+    /// For regions of memory which might overlap, use [`copy`] instead.
+    ///
+    /// `copy_nonoverlapping` is semantically equivalent to C's [`memcpy`].
+    ///
+    /// [`copy`]: ./fn.copy.html
+    /// [`memcpy`]: https://www.gnu.org/software/libc/manual/html_node/Copying-Strings-and-Arrays.html#index-memcpy
     ///
     /// # Safety
     ///
-    /// Beyond requiring that the program must be allowed to access both regions
-    /// of memory, it is Undefined Behavior for source and destination to
-    /// overlap. Care must also be taken with the ownership of `src` and
-    /// `dst`. This method semantically moves the values of `src` into `dst`.
-    /// However it does not drop the contents of `dst`, or prevent the contents
-    /// of `src` from being dropped or used.
+    /// Behavior is undefined if any of the following conditions are violated:
+    ///
+    /// * Both `src` and `dst` must be [valid].
+    ///
+    /// * Both `src` and `dst` must be properly aligned.
+    ///
+    /// * `src.offset(count-1)` must be [valid]. In other words, the region of
+    ///   memory which begins at `src` and has a length of `count *
+    ///   size_of::<T>()` bytes must belong to a single, live allocation.
```

Comment:
The part after "in other words" says way more than the the first part! Also, Why is "`src` must be valid" and "`src.offset(count-1)`" split into two conditions? I see one condition here, which is "`src.offset(i)` must be valid for reading `size_of::<T>()` bytes for `0 <= i < count`", or, equivalently, "`src` must be valid for reading `size_of::<T>() * count` bytes`.

Reply:
I now see in your comments in the thread that you argue it would otherwise violate aliasing rules. That seems plausible but unnecessarily contorted for a definition.

Given that validity has to be defined for a size anyway, I'd just change this to require validity of the entire affected memory range at once (i.e. length `size_of::<T>() * count`).

Reply:
I think I got a bit to clever here :) I was trying to rely exclusively on the definition of validity to define invariants over arrays as well as single accesses. The notion of a "single, live allocation" doesn't appear anywhere in the reference to my knowledge, which is why it went in "in other words".

As I note in my later comment, I think that making pointer validity a function of access size overcomplicates things a bit, and that the type of the pointer should fully specify the number of bytes being accessed when reading or writing.

From this point of view, your `src.offset(i)` for all `i` in `0..count` works formulation nicely, although it still references offset, which is maybe a bit opaque.

Reply:
> As I note in my later comment, I think that making pointer validity a function of access size overcomplicates things a bit, and that the type of the pointer should fully specify the number of bytes being accessed when reading or writing.

In my eyes, making it depend on the whole *type* makes things more complicated. With your proposal, validity is a function of the pointer and its type. A type contains much more information than just a size! With my proposal, it is crystal clear that the size of the only part that actually matters. Plus, for functions where the type alone does not suffice, we don't have to do awkward things like use `.offset` in our definition and rely on that function to be undefined when crossing allocation boundaries.

Reply:
Does this mean I would need to qualify that, for example, `ptr::read` requires that its argument be valid for a read of at least size_of::<T>() bytes? This is what I meant by overcomplicates, not that it wasn't necessary for a formal definition, but it seems redundant when explicitly stated in the documentation.

Simply declaring that when the size of an access for `*const T` is not explicitly stated, we mean valid for `size_of::<T>()` bytes would work.

Reply:
I like this!